platform/x11/cog-platform-x11.c

```diff
@@ -0,0 +1,896 @@
+{
+    xcb_generic_event_t *event = NULL;
+
+    while ((event = xcb_poll_for_event (xcb_data.connection))) {
+        switch (event->response_type & 0x7f) {
+        case XCB_CONFIGURE_NOTIFY:
+        {
+            xcb_configure_notify_event_t *configure_notify = (xcb_configure_notify_event_t *) event;
+            if (configure_notify->width == xcb_data.width
+                && configure_notify->height == xcb_data.height)
+                break;
+
+            xcb_data.width = configure_notify->width;
+            xcb_data.height = configure_notify->height;
+
+            wpe_view_backend_dispatch_set_size (wpe_view_data.backend,
+                                                xcb_data.width,
+                                                xcb_data.height);
+            xcb_schedule_repaint ();
+            break;
+        }
+        case XCB_CLIENT_MESSAGE:
+        {
+            xcb_client_message_event_t *client_message = (xcb_client_message_event_t *) event;
+            if (client_message->window != xcb_data.window)
+                break;
+
+            if (client_message->type == xcb_data.atom_wm_protocols
+                && client_message->data.data32[0] == xcb_data.atom_wm_delete_window) {
+                g_application_quit (g_application_get_default ());
+                break;
+            }
+
+            if (client_message->type == XCB_ATOM_NOTICE) {
+                if (xcb_data.needs_initial_paint) {
+                    xcb_initial_paint ();
+                    xcb_data.needs_initial_paint = false;
+                }
+
+                if (xcb_data.needs_frame_completion) {
+                    xcb_frame_completion ();
+                    xcb_data.needs_frame_completion = false;
+                }
+                break;
+            }
+            break;
+        }
+        case XCB_KEY_PRESS:
+            xcb_handle_key_press ((xcb_key_press_event_t *) event);
+            break;
+        case XCB_KEY_RELEASE:
+            xcb_handle_key_release ((xcb_key_release_event_t *) event);
+            break;
+        case XCB_BUTTON_PRESS:
+            xcb_handle_button_press ((xcb_button_press_event_t *) event);
+            break;
+        case XCB_BUTTON_RELEASE:
+            xcb_handle_button_release ((xcb_button_release_event_t *) event);
+            break;
+        case XCB_MOTION_NOTIFY:
+            xcb_handle_motion_event ((xcb_motion_notify_event_t *) event);
+            break;
+        default:
+            break;
+        }
+    }
+};
+
+struct xcb_source {
+    GSource source;
+    GPollFD pfd;
+};
+
+static gboolean
+xcb_source_check (GSource *base)
+{
+    struct xcb_source *source = (struct xcb_source *) base;
+    return !!source->pfd.revents;
+}
+
+static gboolean
+xcb_source_dispatch (GSource *base, GSourceFunc callback, gpointer user_data)
+{
+    struct xcb_source *source = (struct xcb_source *) base;
+    if (source->pfd.revents & (G_IO_ERR | G_IO_HUP))
+        return FALSE;
```

Comment:
This would be more readable using symbolic constants `G_SOURCE_{CONTINUE,REMOVE}`:

```suggestion
        return G_SOURCE_REMOVE;
```

Also, I am missing here code for handling an errored XCB connection, it would be good to do a check with `xcb_connection_has_error()` and return `G_SOURCE_REMOVE` in that case as well.